From 822e7c2530dc377002253d5ad88540ef78ac8e04 Mon Sep 17 00:00:00 2001
From: Grey Li <grey.li@rock-chips.com>
Date: Wed, 20 Nov 2019 17:20:52 +0800
Subject: [PATCH] video: rockchip: mpp: fix error when mpp_dev_reset

If the power domain of iommu is not change, it will be
return when iommu attach. So there must refresh pm_runtime
for iommu really attach.

Change-Id: I0c8443bd7b9c640c24c23af1693261930144674b
Signed-off-by: Grey Li <grey.li@rock-chips.com>
Signed-off-by: Ding Wei <leo.ding@rock-chips.com>
---
 drivers/video/rockchip/mpp/mpp_common.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/drivers/video/rockchip/mpp/mpp_common.c b/drivers/video/rockchip/mpp/mpp_common.c
index 2b25e6edf0a7..734804939ec7 100644
--- a/drivers/video/rockchip/mpp/mpp_common.c
+++ b/drivers/video/rockchip/mpp/mpp_common.c
@@ -236,16 +236,20 @@ mpp_free_task(struct mpp_session *session,
 	return 0;
 }
 
-static int mpp_flush_pm_runtime(struct device *dev)
+static int mpp_refresh_pm_runtime(struct device *dev)
 {
 	struct device_link *link;
 
-	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node)
-		pm_runtime_get_sync(link->supplier);
+	rcu_read_lock();
 
 	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node)
 		pm_runtime_put_sync(link->supplier);
 
+	list_for_each_entry_rcu(link, &dev->links.suppliers, c_node)
+		pm_runtime_get_sync(link->supplier);
+
+	rcu_read_unlock();
+
 	return 0;
 }
 
@@ -265,12 +269,18 @@ static int mpp_dev_reset(struct mpp_dev *mpp)
 	down_write(&mpp->rw_sem);
 	atomic_set(&mpp->reset_request, 0);
 	mpp_iommu_detach(mpp->iommu_info);
+
 	rockchip_save_qos(mpp->dev);
 	if (mpp->hw_ops->reset)
 		mpp->hw_ops->reset(mpp);
 	rockchip_restore_qos(mpp->dev);
-	/* flush power domain runtime */
-	mpp_flush_pm_runtime(mpp->dev);
+
+	/* Note: if the domain does not change, iommu attach will be return
+	 * as an empty operation. Therefore, force to close and then open,
+	 * will be update the domain. In this way, domain can really attach.
+	 */
+	mpp_refresh_pm_runtime(mpp->dev);
+
 	mpp_iommu_attach(mpp->iommu_info);
 	up_write(&mpp->rw_sem);
 
-- 
2.35.3

